polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs

```diff
@@ -722,34 +763,78 @@ impl FragmentChain {
-			// If they used to be pending availability, don't add them. This is fine
-			// because:
-			// - if they still are pending availability, they have already been added to the new
-			//   storage.
-			// - if they were included, no point in keeping them.
-			//
-			// This cannot happen for the candidates in the unconnected storage. The pending
-			// availability candidates will always be part of the best chain.
-			if self.scope.get_pending_availability(&candidate.candidate_hash).is_none() {
-				let _ = storage.add_candidate_entry(CandidateEntry {
-					candidate_hash: candidate.candidate_hash,
-					parent_head_data_hash: candidate.parent_head_data_hash,
-					output_head_data_hash: candidate.output_head_data_hash,
-					relay_parent: candidate.relay_parent(),
-					candidate: candidate.fragment.candidate_clone(), // This clone is very cheap.
-					state: CandidateState::Backed,
-				});
-			}
+		// Mark the candidate hash.
+		self.unconnected.mark_backed(newly_backed_candidate);
+
+		// Revert to parent_head_hash
+		if !self.revert_to(&parent_head_hash) {
+			// If nothing was reverted, there is nothing we can do for now.
+			return
+		}
+
+		let mut prev_storage = std::mem::take(&mut self.unconnected);
+
+		// Populate the chain.
+		self.populate_chain(&mut prev_storage);
+
+		// Now that we picked the best backable chain, trim the forks generated by candidates
+		// which are not present in the best chain. We can start trimming from this candidate
+		// onwards.
+		self.trim_uneligible_forks(&mut prev_storage, Some(parent_head_hash));
+
+		// Finally, keep any candidates which haven't been trimmed but still have potential.
+		self.populate_unconnected_potential_candidates(prev_storage);
+	}
+
+	/// Checks if this candidate could be added in the future to this chain.
+	/// This will return `Error::CandidateAlreadyKnown` if the candidate is already in the chain or
+	/// the unconnected candidate storage.
+	pub fn can_add_candidate_as_potential(
+		&self,
+		candidate: &impl HypotheticalOrConcreteCandidate,
+	) -> Result<(), Error> {
+		let candidate_hash = candidate.candidate_hash();
+
+		if self.best_chain.contains(&candidate_hash) || self.unconnected.contains(&candidate_hash) {
+			return Err(Error::CandidateAlreadyKnown)
+		}
+
+		self.check_potential(candidate)
+	}
+
+	/// Try adding a seconded candidate, if the candidate has potential. It will never be added to
+	/// the chain directly in the seconded state, it will only be part of the unconnected storage.
+	pub fn try_adding_seconded_candidate(
+		&mut self,
+		candidate: &CandidateEntry,
+	) -> Result<(), Error> {
+		if candidate.state == CandidateState::Backed {
+			return Err(Error::IntroduceBackedCandidate);
 		}
 
-		storage
+		self.can_add_candidate_as_potential(candidate)?;
+
+		// This clone is cheap, as it uses an Arc for the expensive stuff.
+		// We can't consume the candidate because other fragment chains may use it also.
```

Comment:
Best to leave that decision to the caller. If we need a CandidateEntry and not a &CandidateEntry, we should just state it so and leave it to the caller whether any cloning is required or not.

Reply:
the idea was to not clone unless we have to. There's no need to clone if the `self.can_add_candidate_as_potential(candidate)?;` returns an Error